### PR TITLE
Handling input from the data field

### PR DESCRIPTION
**Pull request by szynwelski**

### CODE REVIEW

File: src/utils/smartweave/tags.go [hunk 1]
```diff
@@ -3,15 +3,18 @@ package smartweave
...
-	TagContractSrcTxId = "Contract-Src"
-	TagSDK             = "SDK"
-	TagMinFee          = "Min-Fee"
+	TagAppName              = "App-Name"
+	TagAppNameValue         = "SmartWeaveAction"
+	TagAppVersion           = "App-Version"
+	TagContractTxId         = "Contract"
+	TagInput                = "Input"
+	TagInputFormat          = "Input-Format"
+	TagInputFormatTagValue  = "tag"
```
Comment (janekolszak) [inline src/utils/smartweave/tags.go:12]: Move to another const group, add a comment that those are values of the input-format tag

File: src/utils/smartweave/utils.go [hunk 2]
```diff
@@ -14,16 +14,25 @@ func ValidateInteraction(tx *arweave.Transaction) (isInteraction bool, err error
...
 			if string(tag.Value) == TagAppNameValue {
 				isInteraction = true
 			}
 		case TagInput:
-			// Input tag must be a valid JSON
-			if tool.CheckJSON(tag.Value) != nil {
-				err = fmt.Errorf("value of the tag '%s' is not a valid JSON", TagInput)
+			input = tag.Value
+		case TagInputFormat:
+			if string(tag.Value) == TagInputFormatTagValue {
```
Comment (janekolszak) [inline src/utils/smartweave/utils.go:29]: if you use a switch there will be only one conversion to string

File: src/utils/smartweave/utils.go [hunk 3]
```diff
@@ -39,6 +48,14 @@ func ValidateInteraction(tx *arweave.Transaction) (isInteraction bool, err error
 		return false, nil
 	}
 
+	if !inputFromTag {
+		input = tx.Data
```
[the remark this answers is not in the document]
Comment (janekolszak) [reply src/utils/smartweave/utils.go:52]: You may check data size before downloading

File: src/utils/smartweave/utils.go [hunk 4]
```diff
@@ -49,3 +68,18 @@ func ValidateInteraction(tx *arweave.Transaction) (isInteraction bool, err error
 
 	return true, nil
 }
+
+// whether the interaction should have a non-empty data field
+func InteractionWithData(tx *arweave.Transaction) bool {
```
[the remark this answers is not in the document]
Comment (szynwelski) [reply src/utils/smartweave/utils.go]: done

File: src/utils/smartweave/utils.go [hunk 5]
```diff
@@ -49,3 +68,18 @@ func ValidateInteraction(tx *arweave.Transaction) (isInteraction bool, err error
...
+	if tx == nil || tx.Format < 2 {
+		return false
+	}
+
+	for _, tag := range tx.Tags {
```
[the remark this answers is not in the document]
Comment (szynwelski) [reply src/utils/smartweave/utils.go]: done

File: src/utils/arweave/client.go [hunk 6]
```diff
@@ -313,26 +313,37 @@ func (self *Client) GetTransactionDataById(ctx context.Context, tx *Transaction)
...
+	out, err2 = self.getCachedTransactionDataById(ctx, tx)
+	if err2 == nil {
+		return
+	}
+
+	return out, errors.Join(err, err2)
 }
 
-func (self *Client) GetCachedTransactionDataById(ctx context.Context, tx *Transaction) (out bytes.Buffer, err error) {
+func (self *Client) getCachedTransactionDataById(ctx context.Context, tx *Transaction) (out bytes.Buffer, err error) {
```
[the remark this answers is not in the document]
Comment (szynwelski) [reply src/utils/arweave/client.go]: done

File: src/utils/arweave/client.go [hunk 7]
```diff
@@ -210,7 +210,7 @@ func (self *Client) GetTransactionById(ctx context.Context, id string) (out *Tra
 }
 
 // https://docs.arweave.org/developers/server/http-api#get-transaction-offset-and-size
-func (self *Client) GetTransactionOffsetInfo(ctx context.Context, id string) (out *OffsetInfo, err error) {
+func (self *Client) getTransactionOffsetInfo(ctx context.Context, id string) (out *OffsetInfo, err error) {
```
[the remark this answers is not in the document]
Comment (szynwelski) [reply src/utils/arweave/client.go]: done

File: src/utils/listener/transaction_downloader.go [hunk 8]
```diff
@@ -38,6 +39,8 @@ func NewTransactionDownloader(config *config.Config) (self *TransactionDownloade
 	// No time limit by default
 	self.filter = func(tx *arweave.Transaction) bool { return true }
 
+	self.getTransactionData = func(tx *arweave.Transaction) bool { return false }
```
Comment (janekolszak) [inline src/utils/listener/transaction_downloader.go:42]: isGetTransactionData or anything starting with "is" :)

File: src/utils/arweave/client.go [hunk 9]
```diff
@@ -313,23 +313,34 @@ func (self *Client) GetTransactionDataById(ctx context.Context, tx *Transaction)
...
 	if err != nil {
 		return
 	}
 
-	out.Write(buf)
+	out.Write(data)
 	if out.Len() > 0 {
 		// Do the checks and return
-		if out.Len() != int(tx.DataSize.Int64()) {
+		if len(data) != int(tx.DataSize.Int64()) {
```
Comment (janekolszak) [inline src/utils/arweave/client.go:324]: basically the same as in the original code

File: src/utils/smartweave/utils.go [hunk 10]
```diff
@@ -39,6 +49,18 @@ func ValidateInteraction(tx *arweave.Transaction) (isInteraction bool, err error
 		return false, nil
 	}
 
+	if !inputFromTag {
+		input = tx.Data
+	}
+	// Input must be a valid JSON
+	if jsonError := tool.CheckJSON(input); jsonError != nil {
```
Comment (janekolszak) [inline src/utils/smartweave/utils.go:56]: input may be nil here. It won't validate as json though, so it'll be ok

File: src/utils/smartweave/utils.go [hunk 11]
```diff
@@ -39,6 +49,18 @@ func ValidateInteraction(tx *arweave.Transaction) (isInteraction bool, err error
...
+		err = fmt.Errorf("value of the input is not a valid JSON: %s", jsonError.Error())
+	}
+
+	if tx.Size() > MaxInteractionDataItemSizeBytes {
+		err = fmt.Errorf("the size of the interaction exceeds the limit: %d bytes", MaxInteractionDataItemSizeBytes)
```
Comment (janekolszak) [inline src/utils/smartweave/utils.go:61]: we should probably have dedicated error for this and log those interactions differently. But it's a TODO

File: src/utils/arweave/client.go (hunk 9 above)
Comment (szynwelski) [reply src/utils/arweave/client.go]: this change is the result of a previous refactoring that has been rolled back
I restored the original code

**Review by janekolszak — APPROVED** [review]

**Pull request merged**